### PR TITLE
[libssh2] Fix various issues + add features

### LINKED ISSUE #12939: [libssh2] provide different crypto backends
`libssh2` should provide different crypto backends, similar to existing `libssh` port.

- openssl (default backend)
- wincng (Windows only)
- mbedtls

also make `zlib` feature optional

### PR DESCRIPTION
**Pull request by Thomas1664**
**Describe the pull request**

- #### What does your PR fix?  
  Fixes #12939
  Fix copyright
  Fix deprecated cmake
  ~~Fix issues with DLL not exporting symbols~~

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?  
  Unchanged

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?  
  Yes

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?  
  I am still working on this PR

TODO:

- [x] Make `zlib` optional
- [ ] Support for other crypto libraries (Won't fix, see below)

Reference: https://github.com/libssh2/libssh2/blob/master/docs/INSTALL_CMAKE.md

### CODE REVIEW

File: ports/libssh2/portfile.cmake [hunk 1]
```diff
@@ -4,29 +4,35 @@ vcpkg_from_github(
...
     HEAD_REF master
-    PATCHES "${CMAKE_CURRENT_LIST_DIR}/0001-Fix-UWP.patch"
+    PATCHES "0001-Fix-UWP.patch"
 )
 
-vcpkg_configure_cmake(
+set(OPTIONS "")
+if(VCPKG_LIBRARY_LINKAGE STREQUAL "dynamic")
+    set(OPTIONS "-DBUILD_SHARED_LIBS=ON")
+endif()
```
Comment (JackBoosY) [inline ports/libssh2/portfile.cmake:13]: `BUILD_SHARED_LIBS` was already set in `vcpkg_cmake_configure` according to `VCPKG_LIBRARY_LINKAGE`.
https://github.com/microsoft/vcpkg/blob/b726998b816c583ff549e29d98fe06dbd6d0f694/ports/vcpkg-cmake/vcpkg_cmake_configure.cmake#L263-L271

**Review by JackBoosY — CHANGES_REQUESTED** [review]

File: ports/libssh2/portfile.cmake (hunk 1 above)
Comment (Thomas1664) [reply ports/libssh2/portfile.cmake]: Didn't know why I needed this in the fist place (maybe `vcpkg_configure_cmake` vs `vcpkg_cmake_configure`). Builds now without it.

**Comment by Thomas1664** [comment]
Does implementing alternatives to OpenSSL not violate the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md#do-not-use-features-to-implement-alternatives)?

> Do not use features to implement alternatives
> Features must be treated as additive functionality. If port[featureA] installs and port[featureB] installs, then port[featureA,featureB] must install. Moreover, if a second port depends on [featureA] and a third port depends on [featureB], installing both the second and third ports should have their dependencies satisfied.

**Comment by Thomas1664** [comment]
@JackBoosY There is an [upstream issue](https://github.com/libssh2/libssh2/issues/595) about dll bugs causing our post build checks to fail. Is disabling dynamic linking on Windows and UWP ok?

**Comment by JackBoosY** [comment]
@Thomas1664 Yes we can accept it.

**Comment by JackBoosY** [comment]
> Does implementing alternatives to OpenSSL not violate the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md#do-not-use-features-to-implement-alternatives)?
> 
> > Do not use features to implement alternatives
> > Features must be treated as additive functionality. If port[featureA] installs and port[featureB] installs, then port[featureA,featureB] must install. Moreover, if a second port depends on [featureA] and a third port depends on [featureB], installing both the second and third ports should have their dependencies satisfied.

Sadly, I think so.

**Comment by Thomas1664** [comment]
> Sadly, I think it will.

@JackBoosY Then I'm ready for review

**Comment by JackBoosY** [comment]
I noticed that you left a TODO item, is this PR really ready for review?

File: ports/libssh2/vcpkg.json [hunk 2]
```diff
@@ -1,10 +1,31 @@
 {
   "name": "libssh2",
   "version": "1.10.0",
+  "port-version": 1,
   "description": "libssh2 is a client-side C library implementing the SSH2 protocol.",
   "homepage": "https://www.libssh2.org",
+  "license": "BSD-3-Clause",
+  "supports": "!(!static & (windows | uwp))",
```
Comment (JackBoosY) [inline ports/libssh2/vcpkg.json:8]: So why So why was there no error without these conditions originally? The source code is unchanged.

**Review by JackBoosY — CHANGES_REQUESTED** [review]

File: ports/libssh2/vcpkg.json (hunk 2 above)
Comment (Thomas1664) [reply ports/libssh2/vcpkg.json:8]: The last update of libssh2 was at 11th October last year. Maybe there was a change in CI checks since then?

Comment (JackBoosY) [reply ports/libssh2/vcpkg.json:8]: I will check it locally.

Comment (Thomas1664) [reply ports/libssh2/vcpkg.json:8]: @JackBoosY I figured it out: The new cmake functions are the problem.
From PR #20364:

> One thing I am not sure about is vcpkg build validation errors related to the DLL not having any exports and no import library getting built when I migrated to the new way of including CMake commands (vcpkg-cmake and vcpkg-cmake-config dependencies). Only after reverting this change (https://github.com/microsoft/vcpkg/commit/0fe70113b3b55579fafb0c0b0da3fcaf95377f91) it works again.

So it seems vcpkg_configure_cmake is not equivalent to vcpkg_cmake_configure? Do you know what needs to be changed to make it work with the new commands?

Comment (JackBoosY) [reply ports/libssh2/vcpkg.json:8]: Let me try to compile the dynamic one first,

Comment (Thomas1664) [reply ports/libssh2/vcpkg.json]: Fixed setting `WINDOWS_USE_MSBUILD` in `vcpkg_cmake_configure`

Comment (JackBoosY) [reply ports/libssh2/vcpkg.json]: I believe this changes can not fix this.
In _SOURCE/include/libssh2.h_ line 101:
```cpp
#ifndef LIBSSH2_API
# ifdef LIBSSH2_WIN32
#  ifdef _WINDLL
#   ifdef LIBSSH2_LIBRARY
#    define LIBSSH2_API __declspec(dllexport)
#   else
#    define LIBSSH2_API __declspec(dllimport)
#   endif /* LIBSSH2_LIBRARY */
#  else
#   define LIBSSH2_API
#  endif
# else /* !LIBSSH2_WIN32 */
#  define LIBSSH2_API
# endif /* LIBSSH2_WIN32 */
#endif /* LIBSSH2_API */
```
In dynamic-windows build, `LIBSSH2_WIN32`, `_WINDLL` and `LIBSSH2_LIBRARY` should be defined in the cmake code, but I didn't find the last 2 definitions.

**Comment by Thomas1664** [comment]
@JackBoosY Getting following error un OSX:

```
CMake Error at /Users/vagrant/Data/installed/x64-osx/share/vcpkg-cmake/vcpkg_cmake_configure.cmake:228 (message):
  Unable to determine appropriate generator for: Darwin-x64-external
Call Stack (most recent call first):
  ports/libssh2/portfile.cmake:15 (vcpkg_cmake_configure)
  scripts/ports.cmake:145 (include)
```

**Review by JackBoosY — APPROVED** [review]
LGTM if the tests passed.

**Comment by JonLiu1993** [comment]
The function is successfully tested on the local machine

**Comment by swt2c** [comment]
@Thomas1664, this seems to have caused a regression for curl, see #23505.

**Pull request merged**